apps/ejabberd/src/ejabberd_c2s.erl

```diff
@@ -772,7 +756,7 @@ wait_for_session_or_sm({xmlstreamelement, El}, StateData0) ->
         #iq{type = set, xmlns = ?NS_SESSION} ->
             maybe_open_session(El, StateData);
         _ ->
-            fsm_next_state(wait_for_session_or_sm, StateData)
+            maybe_do_compress(El, wait_for_session_or_sm, StateData)
```

Comment:
I think we should end up in the same state after enabling the compression, but this time compression shouldn't be allowed. How else the binding process will happen?